File: 1325-52-2226-maximum-candies-allocated-to-k-children/1325-52-2226-maximum-candies-allocated-to-k-children.cpp
class Solution {
public:
    inline bool ok(int md,vector<int>& candies, long long k){
        for(int x : candies)
            k-=x/md;
        return  k <= 0; 
    }
    int maximumCandies(vector<int>& candies, long long k) {
        int st = 1, ed = *max_element(candies.begin(),candies.end()),cur = 0,md;
        while(st<=ed){
            md = (st+ed)/2;
            if(ok(md,candies,k)){
                cur = md;
                st = md+1;
            }
            else 
                ed = md-1;
        }
        return cur;
    }
};

File: 1325-52-2226-maximum-candies-allocated-to-k-children/1325-52-2226-maximum-candies-allocated-to-k-children.js
/**
 * @param {number[]} candies
 * @param {number} k
 * @return {number}
 */
var maximumCandies = function(candies, k) {
    const ok = (md) => {
            let cnt = 0;
            for (const x of candies) {
                cnt += Math.floor(x / md);
            }
            return cnt >= k;
        };

        let st = 1;
        let ed = Math.max(...candies);
        let cur = 0;

        while (st <= ed) {
            const md = Math.floor((st + ed) / 2);
            if (ok(md)) {
                cur = md;
                st = md + 1;
            } else {
                ed = md - 1;
            }
        }
        return cur;
    
};